util/fibers/uring_proactor.cc

```diff
@@ -418,45 +412,73 @@ void UringProactor::CancelPeriodicInternal(uint32_t val1, uint32_t val2) {
 }
 
 unsigned UringProactor::RegisterFd(int source_fd) {
-  DCHECK(register_fd_);
+  if (!direct_fd_)
+    return kInvalidDirectFd;
 
-  if (!register_fd_)
-    return source_fd;
-
-  auto next = std::find(register_fds_.begin() + next_free_fd_, register_fds_.end(), -1);
+  // TODO: to create a linked list from free fds.
+  auto next = std::find(register_fds_.begin() + next_free_index_, register_fds_.end(), -1);
   if (next == register_fds_.end()) {
     size_t prev_sz = register_fds_.size();
+    // enlarge direct fds table.
     register_fds_.resize(prev_sz * 2, -1);
```

Comment:
Consider adding some `CHECK(!register_fds_.empty())` (or `DCHECK()`?)

Reply:
Actually, `> 1` given we do `prev_sz + 1` below
(I realize you initialize it to 512, but still, it's conditional and I think it's worth having)

Reply:
it can be 1 so that `next_free_index_` will be 2. it's still a valid case to call `find(end, end, val)`